File: src/java/magma/api/option/Option.java
package magma.api.option;

import magma.api.Tuple2;

import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

public interface Option<T> {
    <R> Option<R> map(Function<T, R> mapper);

    T orElse(T other);

    T orElseGet(Supplier<T> supplier);

    boolean isPresent();

    void ifPresent(Consumer<T> consumer);

    Option<T> or(Supplier<Option<T>> other);

    <R> Option<R> flatMap(Function<T, Option<R>> mapper);

    Option<T> filter(Predicate<T> predicate);

    Tuple2<Boolean,T> toTuple(T other);

    <R> Option<Tuple2<T,R>> and(Supplier<Option<R>> other);
}

File: src/node/magma/api/option/Option.ts
import { Tuple2 } from "../../../magma/api/Tuple2";
export interface Option<T> {map<R>(mapper: (arg0 : T) => R): Option<R>;orElse(other: T): T;orElseGet(supplier: () => T): T;isPresent(): boolean;ifPresent(consumer: (arg0 : T) => void): void;or(other: () => Option<T>): Option<T>;flatMap<R>(mapper: (arg0 : T) => Option<R>): Option<R>;filter(predicate: (arg0 : T) => boolean): Option<T>;toTuple(other: T): Tuple2<boolean, T>;and<R>(other: () => Option<R>): Option<Tuple2<T, R>>;
}
